### DISCUSSION TITLE
PPOCRLabel 启动报错

### DISCUSSION THREAD

**Discussion by vision1v1** [PaddlePaddle/PaddleOCR#12618] [Q&A]
请提供下述完整信息以便快速定位问题/Please provide the following information to quickly locate the problem

- 系统环境/System Environment：Windows 10 企业版 LTSC
- 版本号/Version：release/2.7* Paddle：2.5.2  PaddleOCR：2.7.0.3 问题相关组件/Related components：PPOCRLabel, operands
- 运行指令/Command Code：PPOCRLabel --lang ch --kie True
- 完整报错/Complete Error Message：
[2023/11/03 11:09:02] ppocr WARNING: When args.layout is false, args.ocr is automatically set to false
Traceback (most recent call last):
  File "PPOCRLabel.py", line 2840, in <module>
    sys.exit(main())
  File "PPOCRLabel.py", line 2828, in main
    app, _win = get_main_app(sys.argv)
  File "PPOCRLabel.py", line 2818, in get_main_app
    win = MainWindow(lang=args.lang,
  File "PPOCRLabel.py", line 112, in __init__
    result = self.table_ocr('./data/paddle.png', return_ocr_result_in_table=True)
  File "C:\me\PaddleOCR\.venv\lib\site-packages\paddleocr\paddleocr.py", line 759, in __call__
    res, _ = super().__call__(
  File "..\ppstructure\predict_system.py", line 129, in __call__
    res, table_time_dict = self.table_system(
  File "..\ppstructure\table\predict_table.py", line 86, in __call__
    structure_res, elapse = self._structure(copy.deepcopy(img))
  File "..\ppstructure\table\predict_table.py", line 109, in _structure
    structure_res, elapse = self.table_structurer(copy.deepcopy(img))
  File "..\ppstructure\table\predict_structure.py", line 120, in __call__
    data = transform(data, self.preprocess_op)
  File "..\ppocr\data\imaug\__init__.py", line 56, in transform
    data = op(data)
  File "..\ppocr\data\imaug\operators.py", line 94, in __call__
    img.astype('float32') * self.scale - self.mean) / self.std
ValueError: operands could not be broadcast together with shapes (213,488,4) (1,1,3)

我们提供了AceIssueSolver来帮助你解答问题，你是否想要它来解答(请填写yes/no)?/We provide AceIssueSolver to solve issues, do you want it? (Please write yes/no): no

请尽量不要包含图片在问题中/Please try to not include the image in the issue.

**Comment by 384671834** [comment]
同样的错误，看起来是读取PNG文件的时候出现了第4个channel，但是当限定img为前三个channel后，依旧有op方面的报错

**Comment by chaoshixie** [comment]
我这么改暂时没问题，`paddleocr.py`中的`check_img`函数，可以参考下
```
def check_img(img):
    if isinstance(img, bytes):
        img = img_decode(img)
    if isinstance(img, str):
        # download net image
        if is_link(img):
            download_with_progressbar(img, 'tmp.jpg')
            img = 'tmp.jpg'
        image_file = img
        img, flag_gif, flag_pdf = check_and_read(image_file)
        if not flag_gif and not flag_pdf:
            with open(image_file, 'rb') as f:
                img_str = f.read()
            #     img = img_decode(img_str)
            img = cv2.imdecode(np.fromfile(image_file, dtype=np.uint8), cv2.IMREAD_COLOR)
            # img = cv2.imread(image_file)
            if img is None:
                try:
                    buf = BytesIO()
                    image = BytesIO(img_str)
                    im = Image.open(image)
                    rgb = im.convert('RGB')
                    rgb.save(buf, 'jpeg')
                    buf.seek(0)
                    image_bytes = buf.read()
                    data_base64 = str(base64.b64encode(image_bytes),
                                      encoding="utf-8")
                    image_decode = base64.b64decode(data_base64)
                    img_array = np.frombuffer(image_decode, np.uint8)
                    img = cv2.imdecode(img_array, cv2.IMREAD_COLOR)
                except:
                    logger.error("error in loading image:{}".format(image_file))
                    return None
        if img is None:
            logger.error("error in loading image:{}".format(image_file))
            return None
    if isinstance(img, np.ndarray) and len(img.shape) == 2:
        img = cv2.cvtColor(img, cv2.COLOR_GRAY2BGR)

    return img
```

**Comment by vision1v1** [comment]
我是想启动标注工具，就暂时注掉那两个没用的代码，也不知道后面会不会出现其它问题，目前能启动起来了

**Comment by KrisTopherHao** [comment]
我跟你一样的错误，能否标示一下注释哪个代码啊

**Comment by vision1v1** [comment]
> 我跟你一样的错误，能否标示一下注释哪个代码啊

我代码里是 PPOCRLabel 中 111 112 两行。你可以根据你的报错调用栈修改。

**Comment by vision1v1** [comment]
> 我这么改暂时没问题，`paddleocr.py`中的`check_img`函数，可以参考下
> 
> ```
> def check_img(img):
>     if isinstance(img, bytes):
>         img = img_decode(img)
>     if isinstance(img, str):
>         # download net image
>         if is_link(img):
>             download_with_progressbar(img, 'tmp.jpg')
>             img = 'tmp.jpg'
>         image_file = img
>         img, flag_gif, flag_pdf = check_and_read(image_file)
>         if not flag_gif and not flag_pdf:
>             with open(image_file, 'rb') as f:
>                 img_str = f.read()
>             #     img = img_decode(img_str)
>             img = cv2.imdecode(np.fromfile(image_file, dtype=np.uint8), cv2.IMREAD_COLOR)
>             # img = cv2.imread(image_file)
>             if img is None:
>                 try:
>                     buf = BytesIO()
>                     image = BytesIO(img_str)
>                     im = Image.open(image)
>                     rgb = im.convert('RGB')
>                     rgb.save(buf, 'jpeg')
>                     buf.seek(0)
>                     image_bytes = buf.read()
>                     data_base64 = str(base64.b64encode(image_bytes),
>                                       encoding="utf-8")
>                     image_decode = base64.b64decode(data_base64)
>                     img_array = np.frombuffer(image_decode, np.uint8)
>                     img = cv2.imdecode(img_array, cv2.IMREAD_COLOR)
>                 except:
>                     logger.error("error in loading image:{}".format(image_file))
>                     return None
>         if img is None:
>             logger.error("error in loading image:{}".format(image_file))
>             return None
>     if isinstance(img, np.ndarray) and len(img.shape) == 2:
>         img = cv2.cvtColor(img, cv2.COLOR_GRAY2BGR)
> 
>     return img
> ```

谢谢你的回复了，我还没来得及看，后面我在试试。

**Comment by KrisTopherHao** [comment]
> > 我跟你一样的错误，能否标示一下注释哪个代码啊
> 
> 我代码里是 PPOCRLabel 中 111 112 两行。你可以根据你的报错调用栈修改。

非常感谢
